src/python/pants/goal/run_tracker.py

```diff
@@ -390,3 +413,40 @@ def shutdown_worker_pool(self):
     N.B. This exists only for internal use and to afford for fork()-safe operation in pantsd.
     """
     SubprocPool.shutdown(self._aborted)
+
+  def report_target_info(self, scope, target, keys, val):
+    """Add target information to run_info under target_data.
+
+    Will Recursively construct a nested dict with the keys provided.
```

Comment:
OK, but, to verify that we're both on the same page, the use case you mention as your primary concern currently doesn't work. So it must be fixed (and appropriately tested), or we could just punt on the "list of keys" functionality for this release and add it in a separate change without the pressure of a release.  Since your current use only needs a singleton, you could just pass in a single string key, and then the future change that adds the nesting functionality can test for string vs. list key and act accordingly.